File: 859-design-circular-deque/design-circular-deque.java
class MyCircularDeque {

    int[] deque;
    int front;
    int rear;
    int size;
    int capacity;

    public MyCircularDeque(int k) {
        deque = new int[k];
        capacity = k;
        front = 0;
        rear = -1;
        size = 0;
    }

    public boolean insertFront(int value) {

        if (isFull())
            return false;

        front = (front - 1 + capacity) % capacity;
        deque[front] = value;
        size++;

        if (size == 1)
            rear = front;

        return true;
    }

    public boolean insertLast(int value) {

        if (isFull())
            return false;

        rear = (rear + 1) % capacity;
        deque[rear] = value;
        size++;

        if (size == 1)
            front = rear;

        return true;
    }

    public boolean deleteFront() {

        if (isEmpty())
            return false;

        front = (front + 1) % capacity;
        size--;

        return true;
    }

    public boolean deleteLast() {

        if (isEmpty())
            return false;

        rear = (rear - 1 + capacity) % capacity;
        size--;

        return true;
    }

    public int getFront() {
        if (isEmpty())
            return -1;
        return deque[front];
    }

    public int getRear() {
        if (isEmpty())
            return -1;
        return deque[rear];
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean isFull() {
        return size == capacity;
    }
}

File: 859-design-circular-deque/design-circular-deque.py
class MyCircularDeque:

    def __init__(self, k: int):
        self.deque = [0] * k
        self.capacity = k
        self.front = 0
        self.rear = -1
        self.size = 0

    def insertFront(self, value: int) -> bool:

        if self.isFull():
            return False

        self.front = (self.front - 1) % self.capacity
        self.deque[self.front] = value
        self.size += 1

        if self.size == 1:
            self.rear = self.front

        return True

    def insertLast(self, value: int) -> bool:

        if self.isFull():
            return False

        self.rear = (self.rear + 1) % self.capacity
        self.deque[self.rear] = value
        self.size += 1

        if self.size == 1:
            self.front = self.rear

        return True

    def deleteFront(self) -> bool:

        if self.isEmpty():
            return False

        self.front = (self.front + 1) % self.capacity
        self.size -= 1

        return True

    def deleteLast(self) -> bool:

        if self.isEmpty():
            return False

        self.rear = (self.rear - 1) % self.capacity
        self.size -= 1

        return True

    def getFront(self) -> int:
        if self.isEmpty():
            return -1
        return self.deque[self.front]

    def getRear(self) -> int:
        if self.isEmpty():
            return -1
        return self.deque[self.rear]

    def isEmpty(self) -> bool:
        return self.size == 0

    def isFull(self) -> bool:
        return self.size == self.capacity
        


# Your MyCircularDeque object will be instantiated and called as such:
# obj = MyCircularDeque(k)
# param_1 = obj.insertFront(value)
# param_2 = obj.insertLast(value)
# param_3 = obj.deleteFront()
# param_4 = obj.deleteLast()
# param_5 = obj.getFront()
# param_6 = obj.getRear()
# param_7 = obj.isEmpty()
# param_8 = obj.isFull()